Gutenberg: Enqueue editor assets on Gutenberg action

sirreal opened the pull request:
Currently, we're enqueuing all Gutenberg assets on `init`, which is suboptimal (#10325).

Gutenberg offers [`enqueue_block_assets`](https://github.com/WordPress/gutenberg/blob/de2fab7b8d66eea6c1aeb4a51308d47225fc5df8/lib/client-assets.php#L1059-L1077) [`enqueue_block_editor_assets`](https://github.com/WordPress/gutenberg/blob/de2fab7b8d66eea6c1aeb4a51308d47225fc5df8/lib/client-assets.php#L1608-L1618
). This PR takes advantage of those actions rather than `init`.

Spotted in #10327 
Partial fix for #10325

#### Changes proposed in this Pull Request:

* Split view/editor script and style enqueue to appropriate Gutenberg-specific actions.
* Move growing Gutenberg-specific code out of `Jetpack` class and into its own new `Jetpack_Gutenberg` class.
* Move the `wp-i18n` Jetpack locale setup to editor assets where it's used.

#### Testing instructions:

* Does the Gutenberg editor still work correctly? gutenpack-jn
* Are Gutenberg editor scripts still correctly localized? Check for the `Jetpack_Block_Assets_Base_Url` global in the editor.
* make sure that the [sync handling](https://github.com/Automattic/jetpack/pull/10328/files#diff-b42b8628b3d5d856de376e02baace310R297) that used the moved `Jetpack::is_gutenberg_available()` -> `Jetpack_Gutenberg::is_gutenberg_available()` method continues to work correctly

#### Proposed changelog entry for your changes:

* Use `enqueue_block_assets` and `enqueue_block_editor_assets` for Gutenberg assets.
* Introduce `Jetpack_Gutenberg` class with related methods.

#### Follow up:
- Provide a mechanism to correctly set up `wp-i18n` with the Jetpack locale _exactly once_ when required (#10319).
-  D19535-code

sirreal committed:
Enqueue Gutenberg assets on Gutenberg action

Gutenberg fires `enqueue_block_assets` for this purpose.

sirreal commented on class.jetpack.php:
```diff
@@ -7304,7 +7304,7 @@ public static function load_jetpack_gutenberg() {
 		 *
 		 * @param bool true Whether to load Gutenberg blocks
 		 */
-		if ( ! Jetpack::is_gutenberg_available() || ! apply_filters( 'jetpack_gutenberg', true ) ) {
```
Given that this is a Gutenberg-specific action, the `is_gutenberg_available` check is likely redundant.

sirreal commented:
Adding a note that I believe we'll need to enqueue view assets on `enqueue_block_assets` and remove them from this action:

https://github.com/WordPress/gutenberg/blob/de2fab7b8d66eea6c1aeb4a51308d47225fc5df8/lib/client-assets.php#L1067-L1077

sirreal commented:
Hold off reviews, this needs more work, back to in progress…

sirreal commented:
This should be ready now, it's testing correctly for me.

jeherve reviewed (commented):
Now that we hook `load_jetpack_gutenberg` into `enqueue_block_editor_assets`, should we move the contents of `enqueue_gutenberg_locale` into `load_jetpack_gutenberg` so everything is in one place?

sirreal commented:
@jeherve Thanks for mentioning, I noticed the locale is enqueued on the same action. I don't think we have separate localizations for view and editor, so if any view code needs to be localized, that's probably the _wrong_ action!

I think it makes sense to get the enqueues straightened out in this PR. That includes:
- Gutenberg locale - (is there view/editor split? should there be? is this view or editor specific?)
- View assets -> `enqueue_block_assets` (loaded in editor and view)
- Editor assets -> `enqueue_block_editor_assets` (loaded _only_ in editor)

sirreal commented:
@mtias Could you take a look at what [we're proposing here](https://github.com/Automattic/jetpack/pull/10328#issuecomment-430181346) and confirm it's the right direction?

* `enqueue_block_editor_assets` is mentioned here:
  https://wordpress.org/gutenberg/handbook/extensibility/theme-support/
* This seems to be a description of exactly what we're doing, but uses `init` action. Is this out of date?
  https://wordpress.org/gutenberg/handbook/blocks/writing-your-first-block-type/#enqueuing-block-scripts

jeherve commented:
> Gutenberg locale - (is there view/editor split? should there be? is this view or editor specific?)

@sirreal This comes from our first Gutenberg implementation, in #9560:
https://github.com/Automattic/jetpack/pull/9560/commits/9c660da25b85de48fed6e85afde7d0145b530124

I think this may be something we need to address in https://github.com/Automattic/wp-calypso/issues/27814 as well. cc @tyxla 

> Gutenberg locale - (is there view/editor split? should there be? is this view or editor specific?)

I assume that in some cases we'll have strings that are used in the editor but that are not displayed anywhere on the frontend, like most of the settings in the block sidebars for example.

tyxla commented:
> I think this may be something we need to address in Automattic/wp-calypso#27814 as well. cc @tyxla

@jeherve I'm not sure how this is related - could you please elaborate?

jeherve commented:
@tyxla Sorry, I should have been clearer. This is how we currently load translated strings in Gutenberg:
https://github.com/Automattic/jetpack/commit/9c660da25b85de48fed6e85afde7d0145b530124

As @sirreal pointed out, this does not take some things into consideration, like strings needed in the editor but not on the frontend. All the strings end up in one file.

I think it may be something worth considering when figuring out how strings will be brought over from Calypso to Gutenberg, and enqueued.

tyxla commented:
@jeherve regarding i18n: I don't see a concern with the suggested changes here. Jetpack already passes the translations using `wp.i18n.setLocaleData()`, and Gutenberg will be able to use them, as long as they're in the same format (this is currently in progress, formats are currently different). Am I possibly misunderstanding what we are changing here that will affect i18n?

tyxla reviewed (dismissed):
There's no need to use such an early hook like `init` here, so it's a good change.

This is looking and working good to me 👍 Thanks!

jeherve commented:
> Jetpack already passes the translations using wp.i18n.setLocaleData(), and Gutenberg will be able to use them, as long as they're in the same format (this is currently in progress, formats are currently different). Am I possibly misunderstanding what we are changing here that will affect i18n?

Right, this should work just fine. But as we were discussing above, it may come a point where we don't want to have only one set of translations. 

In any case, we don't have the tools for that right now so I suppose this is off topic here.

sirreal committed:
Separate view and editor Gutenberg asset enqueuing

There are different Gutenberg enqueue actions for editor and other
assets. Use them appropriately with our editor and view assets.

sirreal committed:
Move Gutenberg specific functionality into own class

jeherve reviewed (commented):
This does not appear to completely fix #10325. we are still loading `view.css` and `view.js` on every page, even on posts that do not use Gutenberg at all. This seems less than ideal.

This may be an Gutenberg issue though; I don't know if there is a way to enqueue block-related scripts and styles only on pages where the block is used. I believe this is being discussed here:
https://github.com/WordPress/gutenberg/issues/5445

@mtias Do you have any insights on this for us?

simison commented on sync/class.jetpack-sync-module-posts.php:
```diff
@@ -294,7 +294,7 @@ public function is_gutenberg_meta_box_update() {
 			'1' === $_GET['classic-editor'] &&
 			'1' === $_GET['meta_box'] &&
-			Jetpack::is_gutenberg_available()
+			Jetpack_Gutenberg::is_gutenberg_available()
```
Should we rename this to `Jetpack_Gutenberg::is_available()` ?

sirreal replied:
`is_gutenberg_available` leave no room for confusion. Let's just leave it.

simison commented on class.jetpack-gutenberg.php:
```diff
@@ -0,0 +1,175 @@
+	 */
+	public static function is_gutenberg_available() {
+		return function_exists( 'register_block_type' );
+	}
```
Note to follow and rebase with this PR in case it gets merged first: https://github.com/Automattic/jetpack/pull/10132#discussion_r226252748

sirreal commented:
> This does not appear to completely fix #10325. we are still loading view.css and view.js on every page, even on posts that do not use Gutenberg at all. This seems less than ideal.

@jeherve You're right. I've changed the description to "partial fix for #10325" 🙁

Fixing #10325 really wasn't the motivation for this, and I still think this is a marked improvement and something that can be merged on its own.

lezama commented on class.jetpack-gutenberg.php:
```diff
@@ -0,0 +1,175 @@
+		 *
+		 * @param bool true Whether to load Gutenberg blocks
+		 */
+		if ( ! apply_filters( 'jetpack_gutenberg', true ) ) {
```
```suggestion
		return apply_filters( 'jetpack_gutenberg', true );
```

sirreal replied:
I actually wrote that then changed it 🙂 

Maybe with a cast:

```php
		return (bool) apply_filters( 'jetpack_gutenberg', true );
 ```

lezama replied:
or a `! !` :)

sirreal replied:
✅  d24aaca73dbbf615cede3fdf48cb9c38b9c7ae80

sirreal commented on class.jetpack-gutenberg.php:
```diff
@@ -0,0 +1,172 @@
+		wp_add_inline_script(
+			'wp-i18n',
+			'wp.i18n.setLocaleData( ' . Jetpack::get_i18n_data_json() . ', \'jetpack\' );'
+		);
```
This is a bit surprising to see here because `wp-i18n` is not a dependency of the script that is enqueued here 😕 

In practice, that's fine because this action will always run when the editor action is run (which does enqueue scripts that depend on `wp-i18n`. Effectively, it's the same. Leaving it here means this it won't be enqueued until it's required.

Additionally, I expect the view scripts to depend on `wp-i18n` in the near future (#10319).

Let's leave it as it.

simison replied:
This was added with the VR block.

Asked about these same lines here: https://github.com/Automattic/jetpack/pull/10331#issuecomment-431139029

Leaving sounds good, perhaps we should add a note that they aren't used anywhere (until we actually do use them)?

sirreal replied:
🤔 Ok, I've rethought this.

Until we actually need i18n here, we shouldn't do this. If some other frontend script is requiring `wp-i18n`, we'll be adding our Jetpack locale data without need.

That's good in the sense that it "falls through" to the editor when we depend on it, but bad in the sense that our view doesn't depend on it but we may be adding it unnecessarily to on the frontend.

For now, I'm going to move it to the editor section and add a note in #10319.

tyxla replied:
FWIW, we're relying on this piece in order to be able to reuse Jetpack translations in Gutenberg blocks (cc @ockham)

sirreal commented:
Heads up for reviewers:

I just had a very noisy rebase. I think it's because I moved a bunch of code in one commit and a dependency was added in the middle of the deleted block (`wp-url` #10351 ?)

This is conflict free and ready for review 🙇

sirreal committed:
Only add locale data when necessary

lezama reviewed (dismissed):
❤️

jeherve commented on class.jetpack-gutenberg.php:
```diff
@@ -0,0 +1,173 @@
+
+		$rtl = is_rtl() ? '.rtl' : '';
+
+		/**
```
I would recommend only adding the docblock once. It makes it easier to see that this is not the only occurrence of the filter in the codebase, and it makes our doc parser happy. :)

You can replace that second occurrence by a quick comment like `/** This filter is already documented above */`

sirreal replied:
👍 5d106ecdd5d5d2d2549391083fde01b46a805a1b

jeherve commented on class.jetpack-gutenberg.php:
```diff
@@ -0,0 +1,173 @@
+			$editor_script = "$cdn_base/editor.js";
+			$editor_style  = "$cdn_base/editor$rtl.css";
+
+			/**
```
Same as above.

sirreal replied:
👍 5d106ecdd5d5d2d2549391083fde01b46a805a1b

jeherve reviewed (dismissed):
I like it. I just have one tiny request for our parser. Thank you!

sirreal committed:
Remove duplicate filter documentation

jeherve reviewed (approved):
It seems to work well for me now. 👍
